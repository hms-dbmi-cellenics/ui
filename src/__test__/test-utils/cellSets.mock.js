const mockCellSets = {
  initialLoadPending: false,
  error: false,
  loading: false,
  accessible: true,
  hierarchy: [
    {
      key: 'louvain',
      children: [{ key: 'cluster-a' }, { key: 'cluster-b' }, { key: 'cluster-c' }],
      cellIds: new Set(),
    },
    {
      key: 'sample',
      children: [{ key: 'sample-1' }, { key: 'sample-2' }],
      cellIds: new Set(),
    },
    {
      key: 'scratchpad',
      children: [{ key: 'scratchpad-a' }],
      cellIds: new Set(),
    },
  ],
  properties: {
    'cluster-a': {
      name: 'cluster a',
      key: 'cluster-a',
      cellIds: new Set([0, 1, 6]),
      color: '#01FFFF',
    },
    'cluster-b': {
      name: 'cluster b',
      key: 'cluster-b',
      cellIds: new Set([2, 3]),
      color: '#23FFFF',
    },
    'cluster-c': {
      name: 'cluster c',
      key: 'cluster-c',
      cellIds: new Set([4, 5]),
      color: '#45FFFF',
    },
    'sample-1': {
      name: 'Sample 1',
      key: 'sample-1',
      cellIds: new Set([0, 1, 2]),
      color: '#012FFF',
    },
    'sample-2': {
      name: 'Sample 2',
      key: 'sample-2',
      cellIds: new Set([3, 4, 5]),
      color: '#345FFF',
    },
    'scratchpad-a': {
      cellIds: new Set(['5']),
      key: 'scratchpad-a',
      name: 'New Cluster',
      color: '#5FFFFF',
    },
    louvain: {
      cellIds: new Set(),
      name: 'Louvain clusters',
      key: 'louvain',
      type: 'cellSets',
      rootNode: true,
    },
    sample: {
      cellIds: new Set([6, 7, 8]),
      name: 'Samples',
      key: 'sample',
      type: 'metadataCategorical',
      rootNode: true,
    },
    scratchpad: {
      cellIds: new Set(),
      name: 'Scratchpad',
      key: 'scratchpad',
      type: 'cellSets',
      rootNode: true,
    },
  },
};

export {
  // eslint-disable-next-line import/prefer-default-export
  mockCellSets,
};
