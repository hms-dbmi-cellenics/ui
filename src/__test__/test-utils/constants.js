export default {
  EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperiment',
  CLONED_EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperclone',
  PROJECT_WITH_SAMPLES_UUID: '380c53b4-mock-test-8091-4c82336d6d49',
  PROJECT_WITHOUT_SAMPLE_UUID: '4761594b-mock-test-ba89-c2c812d39bb5',
  EXPERIMENT_NAME: 'Test Experiment',
  PROJECT_ID: 'test24b6-8600-test-mock-63822d2fmock',
  SAMPLE_ID: 'test9188-d682-test-mock-cb6d644cmock',
  API_ENDPOINT: /^http?:\/\/localhost:3000.*$/,
  S3_ENDPOINT: 'http://mock.s3.amazonaws.com',
  // pragma: allowlist secret
  JWT_TOKEN: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6Ikpfake.eyJzdWIiOiIxMjM0NTY3ODkwIiwifakeZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTfakeM5MDIyfQ.fakexwRfakeKKF2QT4fwpMeJf36POk6yJV_adQsfake',
  MOCK_DATETIME: '0000-00-00T00:00:00.000Z',
  ETAG: '380c53b4-mock-etag-8091-4c82336d6d49',
};
