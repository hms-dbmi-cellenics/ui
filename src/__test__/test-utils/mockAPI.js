import _ from 'lodash';

// Default platform data
import cellSetsData from '__test__/data/cell_sets.json';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockDemoExperiments from '__test__/test-utils/mockData/mockDemoExperiments.json';

import fake from '__test__/test-utils/constants';

import {
  responseData,
} from '__test__/test-utils/mockData';

const promiseResponse = (
  response,
  options = {},
) => Promise.resolve(new Response(response, options));

const statusResponse = (code, body) => (
  Promise.resolve({
    status: code,
    body: JSON.stringify(body),
  })
);

const delayedResponse = (response, delay = 10000, options = {}) => new Promise((resolve) => {
  setTimeout(() => {
    resolve(new Response(response, options));
  }, delay);
});

const workerDataResult = (data) => Promise.resolve(_.cloneDeep(data));

const fetchWorkMock = (
  mockedResults,
) => ((experimentId, body) => {
  if (typeof mockedResults[body.name] === 'function') {
    return mockedResults[body.name]();
  }

  return workerDataResult(mockedResults[experimentId]);
});

const generateDefaultMockAPIResponses = (experimentId) => ({
  [`experiments/${experimentId}$`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === experimentId)),
  ),
  [`experiments/${experimentId}/processingConfig$`]: () => promiseResponse(
    JSON.stringify(processingConfigData),
  ),
  [`experiments/${experimentId}/cellSets$`]: () => promiseResponse(
    JSON.stringify(cellSetsData),
  ),
  [`experiments/${experimentId}/backendStatus$`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
  experiments$: () => promiseResponse(
    JSON.stringify(responseData.experiments),
  ),
  [`experiments/${experimentId}/samples$`]: () => promiseResponse(
    JSON.stringify(responseData.samples[0]),
  ),
  '/v2/experiments/examples$': () => promiseResponse(
    JSON.stringify(mockDemoExperiments),
  ),
  'experiments/clone$': () => promiseResponse(
    JSON.stringify(fake.CLONED_EXPERIMENT_ID),
  ),
  [`/v2/workRequest/${experimentId}`]: () => statusResponse(200, 'OK'),
});

const mockAPI = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(Object.keys(apiMapping), (matcherStr) => new RegExp(matcherStr).test(path));

  if (!key) {
    return statusResponse({
      status: 404,
      body: `Path ${path} is undefined`,
    });
  }

  return apiMapping[key](req);
};

export default mockAPI;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  delayedResponse,
  workerDataResult,
  fetchWorkMock,
};
