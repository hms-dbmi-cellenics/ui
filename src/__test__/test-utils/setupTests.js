import '@testing-library/jest-dom/extend-expect';
import preloadAll from 'jest-next-dynamic';

import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('localforage');
jest.mock('utils/pushNotificationMessage');

// This is needed, because the unit tests don't register the Vitessce imports
jest.mock('components/data-exploration/DynamicVitessceWrappers', () => ({
  Scatterplot: () => 'Mocked Scatterplot',
  Heatmap: () => 'Mocked Heatmap',
  Spatial: () => 'Mocked Spatial',
}));

beforeAll(async () => {
  // Add stuff that needs to run once, before all tests
  await preloadAll();
});

beforeEach(async () => {
  // Add stuff that needs to run before each test
});
