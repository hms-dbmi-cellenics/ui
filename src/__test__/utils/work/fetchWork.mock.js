const mockGenesListData = ['A', 'B', 'C', 'D'];

// const mockGeneExpressionData = {
//   A: {
//     rawExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//   },
//   B: {
//     rawExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.702857143,
//       stdev: 2.551115536,
//       expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.702857143,
//       stdev: 2.551115536,
//       expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
//     },
//   },
//   C: {
//     rawExpression: {
//       min: 0,
//       max: 3.4,
//       mean: 1.68,
//       stdev: 2.141525936,
//       expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 3.4,
//       mean: 1.68,
//       stdev: 2.141525936,
//       expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
//     },
//   },
//   D: {
//     rawExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//   },
//   E: { hello: 'world' },
// };

// const mockCacheGet = jest.fn((x) => {
//   const mockCacheContents = {
//     fd3161a878f67ebf54018720cffd6a66: 'A', // pragma: allowlist secret
//     '10250a11679234110a1c260d6fd81d3c': 'B', // pragma: allowlist secret
//     f5c957411a28de68f35e1f5c8a29da7e: 'C', // pragma: allowlist secret
//     '33faa711a94a2028b5bae1778126aec0': 'E', // pragma: allowlist secret
//   };

//   if (x in mockCacheContents) {
//     return mockGeneExpressionData[mockCacheContents[x]];
//   }

//   return null;
// });

const mockCacheGet = jest.fn();
const mockCacheSet = jest.fn();
const mockCacheRemove = jest.fn();

const mockFetchWorkRequest = jest.fn(() => true);

const mockSeekFromS3 = jest.fn();

const mockCacheModule = {
  get: jest.fn((x) => mockCacheGet(x)),
  set: jest.fn((key, val) => mockCacheSet(key, val)),
  _remove: jest.fn((key) => mockCacheRemove(key)),
};

const mockFetchWorkModule = {
  __esModule: true,
  fetchWork: mockFetchWorkRequest,
};

const mockQcPipelineStartDate = '2021-01-01T01:01:01.000Z';

export {
  mockGenesListData, mockCacheGet, mockCacheSet,
  mockFetchWorkRequest, mockSeekFromS3,
  mockQcPipelineStartDate, mockCacheModule, mockFetchWorkModule,
};
