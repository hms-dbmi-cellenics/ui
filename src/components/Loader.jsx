import React from 'react';
import PropTypes from 'prop-types';
import { ClipLoader, BounceLoader } from 'react-spinners';
import { Typography } from 'antd';
import useSWR from 'swr';

import fetchAPI from 'utils/http/fetchAPI';

const { Text } = Typography;

const slowLoad = () => (
  <>
    <div style={{ padding: 25 }}>
      <center>
        <BounceLoader
          size={50}
          color='#8f0b10'
          css={{ display: 'block' }}
        />
      </center>
    </div>
    <p>
      <Text>
        This will take a few minutes...
      </Text>
    </p>
    <p>
      <Text type='secondary'>
        We&apos;re setting up your analysis after a period of inactivity. Please wait.
      </Text>
    </p>
  </>
);

const fastLoad = (message) => (
  <>
    <div style={{ padding: 25 }}>
      <ClipLoader
        size={50}
        color='#8f0b10'
      />
    </div>
    <p>
      <Text>
        {message || "We're getting your data ..."}
      </Text>
    </p>
  </>
);

const Loader = ({ experimentId }) => {
  const { data: workerStatus } = useSWR(
    () => (experimentId ? `/v1/experiments/${experimentId}/backendStatus` : null),
    fetchAPI,
  );

  if (!workerStatus) {
    return (
      <div>
        {fastLoad()}
      </div>
    );
  }

  const { worker: { started, ready } } = workerStatus;
  if (started && ready) {
    return (
      <div>
        {fastLoad()}
      </div>
    );
  }

  return (
    <div>
      {slowLoad()}
    </div>
  );
};

Loader.propTypes = {
  experimentId: PropTypes.string.isRequired,
};

export default Loader;
export { fastLoad, slowLoad };
