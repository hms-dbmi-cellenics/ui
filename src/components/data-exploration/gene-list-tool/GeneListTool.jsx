import React, { useState, useEffect, useRef } from 'react';
import {
  useSelector, useDispatch,
} from 'react-redux';
import _ from 'lodash';
import PropTypes from 'prop-types';
import { v4 as uuidv4 } from 'uuid';
import { geneTableUpdateReason } from 'utils/geneTable/geneTableUpdateReason';
import { loadPaginatedGeneProperties } from 'redux/actions/genes';
import GeneTable from '../generic-gene-table/GeneTable';

const GeneListTool = (props) => {
  const {
    experimentId, width, height, uuid,
  } = props;

  const [tableUuid] = useState(uuid);

  const dispatch = useDispatch();
  const properties = useSelector((state) => state.genes.properties.data);
  const propertiesLoading = useSelector((state) => state.genes.properties.loading);

  const viewState = useSelector((state) => state.genes.properties.views[tableUuid]);
  const error = useSelector((state) => state.genes.properties.views[tableUuid]?.error);
  const componentFetching = useSelector(
    (state) => state.genes.properties.views[tableUuid]?.fetching,
  );
  const tableRowKeys = useSelector((state) => state.genes.properties.views[tableUuid]?.data);
  const total = useSelector((state) => state.genes.properties.views[tableUuid]?.total);

  const initialLoad = useRef(true);

  const PROPERTIES = ['dispersions'];

  // aka tableRows
  const [dataShown, setDataShown] = useState([]);

  const isTableLoading = () => {
    // Load if the view is not yet created (pre-dispatch).
    if (!viewState) {
      return true;
    }

    // Load if the view is actively fetching (post-dispatch, pre-result).
    if (componentFetching) {
      return true;
    }

    // Load if anything is updating the properties we are listening for.
    if (_.intersection(PROPERTIES, propertiesLoading).length > 0) {
      return true;
    }

    return false;
  };

  const columns = [
    {
      fixed: 'left',
      title: 'Dispersion',
      key: 'dispersions',
      sorter: true,
      showSorterTooltip: {
        title: `Dispersion describes how much the variance deviates from the mean.
      Genes with high dispersion have a high level of variation between cells in the dataset.`,
      },
      render: (num) => parseFloat(num.toFixed(3)),
    },
  ];

  const onUpdate = (newState, reason) => {
    // We handle `loading` and `loaded` in the HOC, no need to react to these.
    if (reason === geneTableUpdateReason.loaded || reason === geneTableUpdateReason.loading) {
      return;
    }

    dispatch(loadPaginatedGeneProperties(experimentId, PROPERTIES, tableUuid, newState));
  };

  // When data changes, update rows.
  useEffect(() => {
    if (!tableRowKeys || tableRowKeys.length === 0) {
      return;
    }

    const newRows = [];

    tableRowKeys.forEach((key) => {
      newRows.push({
        gene_names: key,
        dispersions: properties[key].dispersions,
      });
    });

    setDataShown(newRows);

    initialLoad.current = false;
  }, [tableRowKeys]);

  return (
    <GeneTable
      experimentId={experimentId}
      initialTableState={{
        sorter: {
          field: 'dispersions',
          columnKey: 'dispersions',
          order: 'descend',
        },
      }}
      onUpdate={onUpdate}
      columns={columns}
      loading={isTableLoading()}
      error={error || false}
      data={dataShown}
      total={total || 0}
      width={width}
      height={height}
    />
  );
};

GeneListTool.defaultProps = {
  uuid: uuidv4(),
};

GeneListTool.propTypes = {
  experimentId: PropTypes.string.isRequired,
  uuid: PropTypes.string,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
};

export default GeneListTool;
