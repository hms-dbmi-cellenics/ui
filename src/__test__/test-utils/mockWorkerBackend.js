import * as socketConnectionMocks from 'utils/socketConnection';
import * as fetchWorkResponseMocks from 'utils/work/fetchWork';

import SocketMock from 'socket.io-mock';

const socketMock = new SocketMock();

jest.mock('utils/work/fetchWork', () => {
  const mockSeekFromS3 = jest.fn();
  const originalModule = jest.requireActual('utils/work/fetchWork');

  return {
    __esModule: true, // Use it when dealing with esModules
    ...originalModule,
    fetchWork: mockSeekFromS3,
  };
});

jest.mock('aws-amplify', () => ({
  configure: () => ({
    Storage: {
      AWSS3: {
        bucket: 'biomage-originals-test',
      },
    },
  }),
  Storage: {
    get: async (ETag) => `http://mock.s3.amazonaws.com/${ETag}`,
  },
  Auth: {
    federatedSignIn: jest.fn(),
    signOut: jest.fn(),
    currentAuthenticatedUser: jest.fn().mockImplementation(async () => ({
      username: 'mockuser',
      attributes: {
        email: 'mock@user.name',
        name: 'Mocked User',
      },
    })),
    currentSession: () => ({
      getIdToken: (() => ({
        getJwtToken: () => 'fakeJwtToken',
      })),
    }),
  },
  Hub: {
    listen: jest.fn(),
  },
}));

jest.mock('utils/socketConnection', () => {
  const mockEmit = jest.fn();
  const mockOn = jest.fn();
  return {
    __esModule: true,
    default: new Promise((resolve) => {
      resolve({ emit: mockEmit, on: mockOn, id: '1234' });
    }),
    mockEmit,
    mockOn,
  };
});

// To mock worker response, modify the response returned from the API
// Set to null to force fetch from API
fetchWorkResponseMocks.fetchWork.mockImplementation(() => null);

// Set up socket emitter mock
socketConnectionMocks.mockEmit.mockImplementation((workRequestType, requestBody) => {
  const responseBody = { response: { error: false } };

  // After emitting, send reply to listener
  socketMock.socketClient.emit(`WorkResponse-${requestBody.ETag}`, responseBody);
});

// Set up socket listener mock
socketConnectionMocks.mockOn.mockImplementation((channel, f) => {
  socketMock.on(channel, (responseBody) => f(responseBody));
});
