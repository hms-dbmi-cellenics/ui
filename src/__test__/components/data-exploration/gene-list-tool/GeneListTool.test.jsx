import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import preloadAll from 'jest-next-dynamic';
import thunk from 'redux-thunk';
import _ from 'lodash';
import { Empty } from 'antd';
import waitForActions from 'redux-mock-store-await-actions';
import { GENES_PROPERTIES_LOADING, GENES_PROPERTIES_LOADED_PAGINATED } from 'redux/actionTypes/genes';
import { fetchWork } from 'utils/work/fetchWork';
import GeneListTool from 'components/data-exploration/gene-list-tool/GeneListTool';

import Loader from 'components/Loader';

import '__test__/test-utils/setupTests';

jest.mock('utils/getTimeoutForWorkerTask', () => ({
  __esModule: true, // this property makes it work
  default: () => 60,
}));

jest.mock('utils/work/fetchWork', () => ({
  fetchWork: jest.fn(() => new Promise((resolve) => resolve({
    rows: [{
      gene_names: 'R3ALG3N3',
      dispersions: 12.3131,
    }],
    total: 1,
  }))),
}));

const mockStore = configureMockStore([thunk]);
let component;
let store;

const experimentId = '1234';

const TEST_UUID = 'testList';

const backendStatus = {
  [experimentId]: {
    status: {
      pipeline: {
        status: 'SUCCEEDED',
        startDate: '2021-01-01T01:01:01.000Z',
      },
    },
  },
};

const initialState = {
  genes: {
    properties: {
      loading: [],
      data: {
        CEMIP: { dispersions: 3.999991789324 },
        TIMP3: { dispersions: 3.4388329 },
        SMYD3: { dispersions: 3.1273264798 },
        I: { dispersions: 0.08756543 },
        J: { dispersions: 1.352342342 },
        K: { dispersions: 33.423142314 },
      },
      views: {
        [TEST_UUID]: {
          fetching: false,
          error: false,
          total: 4,
          data: ['J', 'I', 'K', 'CEMIP'],
        },
      },
    },
    expression: {
      loading: [],
      error: false,
      data: {},
      views: {},
    },
    selected: [],
  },
  cellInfo: {
    focus: {
      key: null,
      store: null,
    },
  },
  backendStatus,
};

const eventStub = { stopPropagation: () => null };

describe('GeneListTool', () => {
  beforeAll(async () => {
    await preloadAll();
  });

  beforeEach(() => {
    store = mockStore(initialState);

    component = mount(
      <Provider store={store}>
        <GeneListTool experimentId={experimentId} uuid={TEST_UUID} width={100} height={200} />
      </Provider>,
    );
  });

  afterEach(() => {
    component.unmount();
  });

  it('renders correctly', () => {
    const table = component.find('Table');
    const spin = component.find(Loader);
    const genesFilter = component.find('FilterGenes');
    expect(spin.length).toEqual(0);
    expect(table.length).toEqual(1);
    expect(genesFilter.length).toEqual(1);
    expect(table.getElement().props.columns.length).toEqual(3);
    expect(table.getElement().props.columns[0].key).toEqual('lookup');
    expect(table.getElement().props.columns[1].title).toEqual('Gene');
    expect(table.getElement().props.columns[2].title).toEqual('Dispersion');
    expect(table.getElement().props.dataSource.length).toEqual(
      initialState.genes.properties.views[TEST_UUID].data.length,
    );
    expect(table.getElement().props.data.length).toEqual(
      initialState.genes.properties.views[TEST_UUID].data.length,
    );
  });

  it('can sort the gene names in alphabetical order', async () => {
    const newPagination = {
      current: 1,
      pageSize: 4,
      showSizeChanger: true,
      total: 4,
    };

    const newSorter = {
      column: {
        dataIndex: 'gene_names',
        key: 'gene_names',
      },
      render: jest.fn(),
      columnKey: 'gene_names',
      field: 'gene_names',
      order: 'ascend',
    };

    const table = component.find('Table');

    act(() => {
      table.getElement().props.onChange(newPagination, {}, newSorter);
    });

    // Wait for side-effect to propagate (properties loading and loaded).
    await waitForActions(store, [GENES_PROPERTIES_LOADING, GENES_PROPERTIES_LOADED_PAGINATED]);

    expect(fetchWork).toHaveBeenCalledWith(
      experimentId,
      {
        limit: 4,
        name: 'ListGenes',
        offset: 0,
        orderBy: 'gene_names',
        orderDirection: 'ASC',
        selectFields: ['gene_names', 'dispersions'],
      },
      store.getState,
      { timeout: 60 },
    );

    expect(store.getActions()[0]).toMatchSnapshot();
    expect(store.getActions()[1]).toMatchSnapshot();
  });

  it('All `eye` buttons are initially unfocused.', () => {
    const table = component.find('Table');

    table.getElement().props.data.forEach((row) => {
      const lookupComponent = mount(
        <Provider store={store}>
          {row.lookup}
        </Provider>,
      );

      const focusButtonTooltip = lookupComponent.find('FocusButton Tooltip');
      expect(focusButtonTooltip.props().title).toContain('Show');
      expect(focusButtonTooltip.props().title).not.toContain('Hide');

      lookupComponent.unmount();
    });
  });

  it('Clicking one of the `eye` buttons triggers appropriate onChange actions.', () => {
    const table = component.find('Table');

    // Render the appropriate `lookup` component
    const lookupComponent = mount(
      <Provider store={store}>
        {table.getElement().props.data[2].lookup}
      </Provider>,
    );

    // Get focus button
    const button = lookupComponent.find('FocusButton Tooltip Button');

    // Simulate clicking
    button.simulate('click', eventStub);

    expect(store.getActions().length).toEqual(3);
    expect(store.getActions()[2]).toMatchSnapshot();

    lookupComponent.unmount();
  });

  it('Having a focused gene triggers focused view for `eye` button.', () => {
    const FOCUSED_GENE = 'CEMIP';

    // Redefine store from `beforeEach`.
    store = mockStore({
      ...initialState,
      cellInfo: {
        ...initialState.cellInfo,
        focus: {
          ...initialState.cellInfo.focus,
          store: 'genes',
          key: FOCUSED_GENE,
        },
      },
    });

    component = mount(
      <Provider store={store}>
        <GeneListTool experimentId={experimentId} uuid={TEST_UUID} width={100} height={200} />
      </Provider>,
    );

    const table = component.find('Table');

    table.getElement().props.data.forEach((row) => {
      const lookupComponent = mount(
        <Provider store={store}>
          {row.lookup}
        </Provider>,
      );

      const focusButtonTooltip = lookupComponent.find('FocusButton Tooltip');

      if (row.gene_names === FOCUSED_GENE) {
        expect(focusButtonTooltip.props().title).toContain('Hide');
      } else {
        expect(focusButtonTooltip.props().title).toContain('Show');
      }

      lookupComponent.unmount();
    });
  });

  it('error state renders correctly', () => {
    const newInitialState = _.cloneDeep(initialState);
    newInitialState.genes.properties.views[TEST_UUID].error = 'asd';
    store = mockStore(newInitialState);

    component = mount(
      <Provider store={store}>
        <GeneListTool experimentId={experimentId} uuid={TEST_UUID} width={100} height={200} />
      </Provider>,
    );

    expect(component.find(Empty).length).toEqual(1);
  });
});
