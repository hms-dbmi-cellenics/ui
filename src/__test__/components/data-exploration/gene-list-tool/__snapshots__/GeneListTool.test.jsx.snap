// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeneListTool Clicking one of the \`eye\` buttons triggers appropriate onChange actions. 1`] = `
Object {
  "payload": Object {
    "experimentId": "1234",
    "key": "J",
    "store": "genes",
  },
  "type": "cellInfo/focus",
}
`;
